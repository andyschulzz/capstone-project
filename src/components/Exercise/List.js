import React from 'react'
import Exercise from './Exercise'
import SearchBox from '../common/SearchBox'
import { Button } from '../common/Button'
import { useRouteMatch, Link } from 'react-router-dom'
import { useForm } from 'react-hook-form'
import * as S from './List.styles'
import PropTypes from 'prop-types'

List.propTypes = {
  exercises: PropTypes.array.isRequired,
  handleExerciseSelect: PropTypes.func.isRequired,
  handleSearch: PropTypes.func.isRequired,
  search: PropTypes.string,
}
const lettersAtoZ = [...Array(26)].map((_, i) =>
  String.fromCharCode('A'.charCodeAt(0) + i)
)

export default function List({
  exercises,
  handleExerciseSelect,
  handleSearch,
  search,
}) {
  const { url } = useRouteMatch()
  const { register, handleSubmit, reset } = useForm({ mode: 'onChange' })

  const renderedExercises = lettersAtoZ.reduce((acc, letter) => {
    const letterExercises = filterExercises(letter)
    if (letterExercises.length !== 0) {
      acc.push(<S.Span key={letter}>{letter}</S.Span>)
      letterExercises.forEach(lc => acc.push(lc))
    }
    return acc
  }, [])
  return (
    <>
      <S.ButtonWrapper>
        <Button primary="true" as={Link} to={`${url}/add`}>
          Add
        </Button>
        <SearchBox
          register={register}
          onSubmit={handleSubmit(onSubmit)}
          reset={reset}
          search={search}
        />
      </S.ButtonWrapper>
      <S.Wrapper>
        {exercises.length ? (
          renderedExercises
        ) : (
          <S.Container>No exercise found!</S.Container>
        )}
      </S.Wrapper>
    </>
  )

  function filterExercises(letter) {
    const filter = exercises
      .filter(exercise => String(exercise.name).startsWith(letter))
      .map((exercise, index) => {
        return (
          <Exercise
            key={exercise.id}
            {...exercise}
            index={index}
            handleExerciseSelect={handleExerciseSelect}
          />
        )
      })
    return filter
  }

  function onSubmit(userInput) {
    handleSearch(userInput)
  }
}
