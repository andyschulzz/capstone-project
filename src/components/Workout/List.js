import React from 'react'
import * as S from './List.styles'

export default function List({ workouts }) {
  const groupByTitle = workouts.reduce((acc, obj) => {
    const key = obj['title']
    if (!acc[key]) {
      acc[key] = []
    }
    acc[key].push(obj)
    return acc
  }, {})

  const renderWorkouts = Object.entries(groupByTitle).map(
    ([title, exercises]) => (
      <React.Fragment key={title}>
        <h3>{title}</h3>
        {exercises.map((exercise, index) => {
          return (
            <S.Wrapper key={index}>
              {exercise.name}
              <S.DetailsWrapper>
                {exercise.reps}
                <S.Span>&times;</S.Span>
                {exercise.sets}
                {exercise.weight && (
                  <S.Span isWeight>({exercise.weight} kg)</S.Span>
                )}
              </S.DetailsWrapper>
            </S.Wrapper>
          )
        })}
      </React.Fragment>
    )
  )
  return <>{renderWorkouts}</>
}
