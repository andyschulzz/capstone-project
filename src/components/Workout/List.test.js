import React from 'react'
import { render, cleanup, fireEvent } from '@testing-library/react'
import List from './List.js'
import { MemoryRouter as Router } from 'react-router-dom'

afterEach(cleanup)
const handleWorkoutDeleteMock = jest.fn()
const handleWorkoutEditMock = jest.fn()
const workoutsMock = [
  {
    id: 1,
    name: 'Test',
    type: 'Test',
    instructions: 'Test',
    title: 'Test',
  },
]

describe('<List />', () => {
  it('should render correctly', () => {
    const { asFragment } = render(
      <Router>
        <List
          workouts={workoutsMock}
          handleWorkoutDelete={handleWorkoutDeleteMock}
          handleWorkoutEdit={handleWorkoutEditMock}
        />
      </Router>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('Workout menu should be visible when button is clicked', () => {
    const { getByTestId } = render(
      <Router>
        <List
          workouts={workoutsMock}
          handleWorkoutDelete={handleWorkoutDeleteMock}
          handleWorkoutEdit={handleWorkoutEditMock}
        />
      </Router>
    )
    fireEvent.click(getByTestId('menu-icon'))
    expect(getByTestId('workout-menu')).toBeVisible()
  })
})
