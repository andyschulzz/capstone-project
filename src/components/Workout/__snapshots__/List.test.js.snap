// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> should render correctly 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-size: 1.1rem;
  margin: 5px 0;
}

.c5 {
  font-size: 1.1rem;
  margin: 0 5px;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  font-size: 0.9rem;
  color: #999;
}

.c2 {
  font-size: 28px;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  margin-left: auto;
  margin-top: -10px;
  color: hsl(200,100%,50%);
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 {
  padding: 20px;
  border: 1px solid #ccc;
  border-radius: 5px;
  position: relative;
}

.c0 h3 {
  margin: 0;
  font-size: 1.4rem;
}

.c0 span {
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <h3 />
      <svg
        class="c2"
        data-testid="menu-icon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
        />
      </svg>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <span
          class="c5"
        >
          ×
        </span>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <span
          class="c5"
        >
          ×
        </span>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <span
          class="c5"
        >
          ×
        </span>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <span
          class="c5"
        >
          ×
        </span>
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
      >
        <span
          class="c5"
        >
          ×
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
